### ISSUE TITLE
[BUG] RangeError Maximum call stack size exceeded during disconnect

### ISSUE THREAD

**Issue by jamiemaertens** [twilio/twilio-voice.js#107]
- [X] I have verified that the issue occurs with the latest twilio.js release and is not marked as a known issue in the [CHANGELOG.md](https://github.com/twilio/twilio-voice.js/blob/master/CHANGELOG.md).
 - [X] I reviewed the [Common Issues](https://github.com/twilio/twilio-voice.js/blob/master/COMMON_ISSUES.md) and open GitHub issues and verified that this report represents a potentially new issue.
 - [X] I verified that the [Quickstart](https://github.com/TwilioDevEd/voice-javascript-sdk-quickstart-node) application works in my environment.
 - [X] I am not sharing any [Personally Identifiable Information (PII)](https://www.twilio.com/docs/glossary/what-is-personally-identifiable-information-pii)
       or sensitive account information (API keys, credentials, etc.) when reporting this issue.

**Code to reproduce the issue:**

```js

  /**
   * Disconnect call and execute modal completion
   */
  const endCall = useCallback(() => {
    connectionRef.current?.disconnect()
    connectionRef.current?.removeAllListeners()
    
    setShowEndCallForm(true)
  }, [])

  // Fetch token only once and connect the call
  useEffect(() => {
    fetchTwilioAccessToken(initialNumber, extension).then(async (data) => {
      const payload = data.payload
      phoneEventId.current = data.phoneEventId
      setPhoneFromNumber(Format.phone(data.fromPhoneNumber, true))
      const twilioDevice = new Device(data.accessToken, {
        sounds: {
          outgoing: audioUrl,
        },
      })

      // Connect the call with the payload when device is ready (which is triggered after setup)
      connectionRef.current = await twilioDevice.connect({ params: { number: initialNumber, payload } })

      // Disconnect call on error or disconnection
      connectionRef.current.on('disconnect', () => {
        endCall()
      })

      connectionRef.current.on('error', (error) => {
        console.info(`Error occurred during twilio call: ${JSON.stringify(error)}`)
        endCall()
      })
    })
  }, [endCall, initialNumber, extension])


```

**Expected behavior:**

* Disconnect works properly

**Actual behavior:**

ran into an issue when executing the disconnect function:

RangeError
Maximum call stack size exceeded

<img width="845" alt="image" src="https://user-images.githubusercontent.com/5513861/185502792-a4b9eff9-1d38-464e-b8ab-e56b38cad2af.png">

Also there was a loop of this error shown in the logs:
```
Error occurred during twilio call: {"code":31009,"message":"No transport available to send or receive messages","twilioError":{"causes":[],"code":31009,"description":"Transport error","explanation":"No transport available to send or receive messages","solutions":[]}}
```

* Please advise if my error handling is incorrect. I fear that calling disconnect within the error handling may be causing a loop.
* Related support ticket: 9290565

**Software versions:**

 - [X] Browser(s): Chrome 102.0.0
 - [X] Operating System: Mac OS X 10.15.7
 - [X] twilio.js:  "@twilio/voice-sdk": "2.1.1"
 - [X] Third-party libraries (e.g., Angular, React, etc.): React: "react": "^17.0.2"

**Comment by bld010** [comment]
Hi @jamiemaertens , Twilio Developer Educator here. 

When the `Call` instance emits the `disconnect` event, the call has already ended. I'm guessing that the error you are seeing is because your `endCall` method invokes `.disconnect()`. The call has already ended, but the SDK is trying to send a `hangup` message to the other party. This causes another `error` event to be emitted, which then invokes that `endCall` method again, creating a loop. 

The event listeners should also be cleaned up by the SDK when a call is disconnected, so unless you're adding your own listeners, you probably don't need to invoke `removeAllListeners()`. 

I would try invoking your `setShowEndCallForm(true)` instead of the `endCall` method in the `error` and `disconnect` handlers.

Hopefully that helps, but if not just let me know!

**Comment by jamiemaertens** [comment]
Hi @bld010 , thanks for your update. I see in your example code that `removeAllListeners` is executed within the disconnect handler:
https://github.com/twilio-labs/dev-phone/blob/main/packages/dev-phone-ui/src/components/WebsocketManagers/VoiceManager.jsx#L65
is that the correct approach?
I am attempting to remove the 'disconnect' and 'error' listeners to ensure they are cleaned up.
I updated my system to not call the `endCall` during processing of those listeners and that worked out well. Thanks!

**Comment by bld010** [comment]
Yes, it looks like those two particular listeners aren't removed by the SDK. You can call removeAllListeners() and that shouldn't hurt anything!

**Comment by jamiemaertens** [comment]
Thanks for the help, I'm all set!
